index.php

```diff
@@ -10,10 +10,17 @@
     ],
     'hooks' => [
         'page.render:after' => function ($contentType, $data, $html) {
-            if ($contentType === 'html') {
-                $html = str_replace('</body>', snippet('admin-bar', [], true), $html) . '</body>';
+            if ($contentType !== 'html') return;
+
+            // find last occurrence of end tag
+            $pos = strripos($html, '</body>');
+            if ($pos === false) $pos = strripos($html, '</html>');
+
+            // insert snippet before end tag
+            if ($pos !== false) {
+                $bar = snippet('admin-bar', [], true);
+                return substr_replace($html, $bar, $pos, 0);
```

Comment:
Using `substr_replace` with the position of the occurrence of `</body>` (or `</html>`) that we found, and a length of `0`, means that we are inserting the replacement string at the desired position but not replacing any character in the original string.

This makes the issue in #3 impossible.